pkg/server/service/loadbalancer/wrr/wrr.go

```diff
@@ -37,79 +44,70 @@ func New(sticky *dynamic.Sticky) *Balancer {
+	if n[i].deadline == n[j].deadline {
+		return n[i].index < n[j].index
 	}
-	return max
+	return n[i].deadline < n[j].deadline
 }
 
-func (b *Balancer) weightGcd() int {
-	divisor := -1
-	for _, s := range b.handlers {
-		if divisor == -1 {
-			divisor = s.weight
-		} else {
-			divisor = gcd(divisor, s.weight)
-		}
-	}
-	return divisor
+// Swap implements heap.Interface/sort.Interface
+func (n namedHandlers) Swap(i, j int) {
+	n[i], n[j] = n[j], n[i]
 }
 
-func gcd(a, b int) int {
-	for b != 0 {
-		a, b = b, a%b
-	}
-	return a
+// Push implements heap.Interface for pushing an item into the heap
+func (n *namedHandlers) Push(x interface{}) {
+	h := x.(*namedHandler)
+	*n = append(*n, h)
+}
+
+// Pop implements heap.Interface for poping an item from the heap
+func (n *namedHandlers) Pop() interface{} {
+	old := *n
+	h := old[len(old)-1]
+	*n = old[0 : len(old)-1]
+	return h
+}
+
+// Balancer is a WeightedRoundRobin load balancer based on Earliest Deadline First (EDF).
+// (https://en.wikipedia.org/wiki/Earliest_deadline_first_scheduling)
+// Each pick from the schedule has the earliest deadline entry selected. Entries have deadlines set
+// at current time + 1 / weight, providing weighted round robin behavior with floating point
+// weights and an O(log n) pick time.
+type Balancer struct {
+	handlers     *namedHandlers
+	mutex        *sync.Mutex
+	curIndex     int64
+	curDeadline  float64
+	baseDeadline float64 // for fixing starting bias
+	stickyCookie *stickyCookie
 }
 
 func (b *Balancer) nextServer() (*namedHandler, error) {
 	b.mutex.Lock()
 	defer b.mutex.Unlock()
-
-	if len(b.handlers) == 0 {
+	if b.handlers == nil || len(*b.handlers) == 0 {
 		return nil, fmt.Errorf("no servers in the pool")
 	}
-
-	// The algo below may look messy, but is actually very simple
-	// it calculates the GCD  and subtracts it on every iteration, what interleaves servers
-	// and allows us not to build an iterator every time we readjust weights
-
-	// GCD across all enabled servers
-	gcd := b.weightGcd()
-	// Maximum weight across all enabled servers
-	max := b.maxWeight()
-
-	for {
-		b.index = (b.index + 1) % len(b.handlers)
-		if b.index == 0 {
-			b.currentWeight -= gcd
-			if b.currentWeight <= 0 {
-				b.currentWeight = max
-				if b.currentWeight == 0 {
-					return nil, fmt.Errorf("all servers have 0 weight")
-				}
-			}
-		}
-		srv := b.handlers[b.index]
-		if srv.weight >= b.currentWeight {
-			log.WithoutContext().Debugf("Service Select: %s", srv.name)
-			return srv, nil
-		}
-	}
+	// Pick handler with closest deadline.
+	handler := heap.Pop(b.handlers).(*namedHandler)
+	// curDeadline should be handler's deadline so that new added entry would have a fair
+	// competition environment with the old ones.
+	b.curDeadline = handler.deadline
+	handler.deadline += 1 / handler.weight
+	b.curIndex++
+	handler.index = b.curIndex
+	heap.Push(b.handlers, handler)
+	log.WithoutContext().Debugf("Service Select: %s", handler.name)
```

Comment:
let's update this log message, which wasn't great to begin with. how about:
("Service selected by WRR: %s", handler.name)
?

Reply:
Thanks, good suggestion.
fixed.